koku/masu/util/common.py

```diff
@@ -570,3 +571,28 @@ def convert_account(account):
 def filter_dictionary(dictionary, keys_to_keep):
     """Filter a dictionary to only include the keys specified."""
     return {key: value for key, value in dictionary.items() if key in keys_to_keep}
+
+
+class SingletonMeta(type):
+    """
+    This is a thread-safe implementation of Singleton.
+    """
+
+    _instances = {}
+    _lock: Lock = Lock()
+
+    def __call__(cls, *args, **kwargs):
+        """
+        Possible changes to the value of the `__init__` argument do not affect
+        the returned instance.
+        """
+        with cls._lock:
+            # The first thread to acquire the lock, reaches this conditional,
+            # goes inside and creates the Singleton instance. Once it leaves the
+            # lock block, a thread that might have been waiting for the lock
+            # release may then enter this section. But since the Singleton field
+            # is already initialized, the thread won't create a new object.
+            if cls not in cls._instances:
+                instance = super().__call__(*args, **kwargs)
+                cls._instances[cls] = instance
+        return cls._instances[cls]
```

Comment:
I've been scrutinizing the idea of using a class var to store the singleton instance and I think there may be an issue with the example I shared earlier: I believe `cls` is the `__call__()` is the class using the metaclass, not the metaclass itself. It seems to work, but I need to look at why a bit more.

Reply:
I think it may be clearer to use an `__init__` in the metaclass and explicitly store the singleton instance of a class on the class itself, not the singleton metaclass. I'll come up with an example.

Reply:
We really need to wrap this PR up. Do we need this singleton class? I'd argue since this is the first time we've used a singleton, let's just use the global variable since it was working. If we need to add a new singleton in the future, then maybe we can implement this metaclass and update things here, but I don't see the value in adding it now since all we are doing is delaying this PR.

Reply:
Using a Singleton is the right move here. Using a global is not thread safe and is susceptible to mutation whereas a Singleton is not.

I worked out what this implementation is doing and it's storing `_instances` on the subclass object itself rather than in the Singleton class attribute. That's fine but not obvious.

I was thinking it would read better and make more sense to store the single on the subclass singleton instance itself, but I don't want to hold up the PR for that. The current implementation works just fine (but `RLock` should be used instead of `Lock`).

Reply:
Here's an example (for future reference or just science) of storing the singleton instance on the instance instead of in a dict in the metaclass object.

```python
# This is @sivel's implementation. Took me a while to get my head around it.
class Singleton(type):
    def __init__(cls, name, bases, dct):
        super().__init__(name, bases, dct)
        cls.__instance = None
        cls.__rlock = RLock()

    def __call__(cls, *args, **kwargs):
        if cls.__instance is not None:
            return cls.__instance

        if cls.__instance is None:
            with cls.__rlock:
                cls.__instance = super().__call__(*args, **kwargs)

        return cls.__instance
```

Name mangling keeps those on the object instance but tucked away.

```
In [7]: msc = MySingletonClass()

In [8]: msc._Singleton__instance
Out[8]: <__main__.MySingletonClass at 0x1082b2670>

In [9]: msc._Singleton__rlock
Out[9]: <unlocked _thread.RLock object owner=0 count=0 at 0x1053d49c0>

In [11]: msc is MySingletonClass()
Out[11]: True
```

This is more about organization than being better or worse than the existing implementation. The only minor difference is that each singleton object instance get its own `RLock` instance.